### PR TITLE
Early return sniff

### PR DESCRIPTION
**Pull request by carusogabriel**
Based on https://github.com/doctrine/doctrine2/pull/6929, https://github.com/doctrine/mongodb-odm/pull/1691, https://github.com/doctrine/doctrine2/pull/6889 and https://github.com/doctrine/dbal/pull/2937

### CODE REVIEW

**Comment by carusogabriel** [comment]
@kukulich Can I ask your help here? Am I missing something to get this to work?

**Comment by kukulich** [comment]
@carusogabriel Hmm, in yesterday perspective I thought that the code should not be reported because there's already "early exit" code in `if`. It looks different to me in today perspective so I implemented it :) Try new build please.

**Comment by kukulich** [comment]
@carusogabriel Just for your info, there are three types of errors in this sniff: "Use early exit instead of else.", "Remove useless else to reduce code nesting." and "Use early exit to reduce code nesting.". It may be good to document and test all of them here.

**Comment by carusogabriel** [comment]
@kukulich Thanks, I'll add more test so

**Comment by kukulich** [comment]
https://github.com/slevomat/coding-standard/releases/tag/4.4.0

**Comment by carusogabriel** [comment]
@kukulich Thanks. Time to update some tests and untag WIP :construction_worker_man:

**Comment by carusogabriel** [comment]
_Solving conflicts_

File: README.md [hunk 1]
```diff
@@ -31,6 +31,7 @@ exceptions/differences/extensions (:white_check_mark: are the implemented sniffs
 - :white_check_mark: Omit phpDoc for parameters/returns with native types, unless adding description
 - :white_check_mark: Don't use `@author`, `@since` and similar annotations that duplicate Git information
 - :white_check_mark: Assignment in condition is not allowed
+- :white_check_mark: Use early return instead of `else`
```
Comment (kukulich) [inline README.md]: The comment is not true. The sniff does a little more :)

File: lib/Doctrine/ruleset.xml [hunk 2]
```diff
@@ -93,6 +93,8 @@
     <rule ref="SlevomatCodingStandard.ControlStructures.DisallowYodaComparison"/>
     <!-- Forbid weak comparisons -->
     <rule ref="SlevomatCodingStandard.ControlStructures.DisallowEqualOperators"/>
+    <!-- Require usage of early exit instead of using else -->
```
Comment (kukulich) [inline lib/Doctrine/ruleset.xml]: @carusogabriel The comment is not true. The sniff does a little more :)

File: README.md (hunk 1 above)
Comment (carusogabriel) [reply README.md]: Have a suggestion? :smile:

File: lib/Doctrine/ruleset.xml (hunk 2 above)
Comment (carusogabriel) [reply lib/Doctrine/ruleset.xml]: Have a suggestion? :smile:

Comment (kukulich) [reply lib/Doctrine/ruleset.xml]: https://github.com/slevomat/coding-standard/commit/4dfc4ed2c43a97983079bbb1d6f2a36833aceb1e 😄

**Comment by Ocramius** [comment]
@carusogabriel https://travis-ci.org/doctrine/coding-standard/jobs/337937694 failed :-\

**Comment by kukulich** [comment]
@carusogabriel There's bad indentation here: https://github.com/doctrine/coding-standard/pull/22/files#diff-50034f4d68d4e08b523b28d23715cd5fR34

And nothing should change in method `quoox`. I think it's ok as it is.

**Comment by carusogabriel** [comment]
@Ocramius @kukulich Thanks, as my editor dealing with tabs :man_facepalming:

**Comment by carusogabriel** [comment]
One more conflict :fire: :fire: :fire: :fire: :fire: :fire: :fire: :fire:

**Comment by Ocramius** [comment]
@carusogabriel yeah, still some weirdness going on there :P

**Comment by Ocramius** [comment]
Rebase also needed, sorry @carusogabriel!

File: tests/fixed/EarlyReturn.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    public function quoox() : bool
+    {
+        if (true !== 'true') {
+            return false;
+        }
+        
+        if (true === 'true') {
+            if (false === false) {
+                return true;
+            }
```
Comment (carusogabriel) [inline tests/fixed/EarlyReturn.php]: @kukulich Something is going crazy here :fire: 

Any clue?

**Comment by Ocramius** [comment]
@carusogabriel thanks! Proceeding to release :)

**Comment by carusogabriel** [comment]
@Ocramius Perfect. Should we update across Doctrine's repos as well?

**Comment by Ocramius** [comment]
@carusogabriel yeah, let's do that afterwards 👍

**Pull request merged**